services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/CloudMailInManagerImpl.java

```diff
@@ -0,0 +1,147 @@
+					}
+				} else if (FROM_HEADER.equalsIgnoreCase(key)) {
+					from = headers.getString(key);
+				}
+			}
+			if (from==null) throw new IllegalArgumentException("Sender ('From') is required.");
+			if (to.isEmpty()) throw new IllegalArgumentException("There must be at least one recipient.");
+			MessageToUser mtu = new MessageToUser();
+			mtu.setSubject(subject);
+			Set<String> recipients = new HashSet<String>();
+			for (String email : to) {
+				recipients.add(lookupPrincipalIdForSynapseEmailAddress(email).toString());
+			}
+			mtu.setRecipients(recipients);
+			mtu.setCreatedBy(lookupPrincipalIdForRegisteredEmailAddress(from).toString());		
+			mtu.setNotificationUnsubscribeEndpoint(notificationUnsubscribeEndpoint);
+			MessageToUserAndBody result = new MessageToUserAndBody();
+			result.setMetadata(mtu);
+			result.setMimeType(ContentType.APPLICATION_JSON.getMimeType());
+			MessageBody messageBody = copyMessageToMessageBody(message);
+			result.setBody(EntityFactory.createJSONStringForEntity(messageBody));
+			return result;
+		} catch (JSONException e) {
+			throw new RuntimeException(e);
+		} catch (JSONObjectAdapterException e) {
+			throw new RuntimeException(e);
+		}
+	}
+	
+	public static MessageBody copyMessageToMessageBody(Message message) {
+		MessageBody result = new MessageBody();
+		// Note, if this is a reply we simply take the reply field and drop the html and plain fields
+		if (message.getReply_plain()!=null && message.getReply_plain().length()>0) {
+			result.setPlain(message.getReply_plain());
+		} else {
+			result.setPlain(message.getPlain());
+			result.setHtml(message.getHtml());
+		}
+		List<Attachment> attachments = new ArrayList<Attachment>();
+		if (message.getAttachments()!=null) {
+			for (org.sagebionetworks.repo.model.message.cloudmailin.Attachment cloudMailInAttachment : 
+					message.getAttachments()) {
+				Attachment attachment = new Attachment();
+				attachment.setContent(cloudMailInAttachment.getContent());
+				attachment.setContent_id(cloudMailInAttachment.getContent_id());
+				attachment.setContent_type(cloudMailInAttachment.getContent_type());
+				attachment.setDisposition(cloudMailInAttachment.getDisposition());
+				attachment.setFile_name(cloudMailInAttachment.getFile_name());
+				attachment.setSize(cloudMailInAttachment.getSize());
+				attachment.setUrl(cloudMailInAttachment.getUrl());
+				attachments.add(attachment);
+			}
+		}
+		result.setAttachments(attachments);
+		return result;
+	}
+
+	public Long lookupPrincipalIdForSynapseEmailAddress(String email) {
+		// first, make sure it's actually an email address
+		AliasEnum.USER_EMAIL.validateAlias(email);
+		String emailLowerCase = email.toLowerCase();
+		if (!emailLowerCase.endsWith(EMAIL_SUFFIX_LOWER_CASE))
+			throw new IllegalArgumentException("Email must end with "+EMAIL_SUFFIX_LOWER_CASE);
+		String aliasString = emailLowerCase.substring(0,  
+				emailLowerCase.length()-EMAIL_SUFFIX_LOWER_CASE.length());
```

Comment:
if (aliasString.equals("noreply")) throw new IllegalArgumentException("You are trying to reply to a 'noreply' email address.") (or something)